File: rust/cudf-sys/cpp/filling.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/filling.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/filling.hpp>

namespace cudf_sys {

std::unique_ptr<Column> fill_column(
    cudf::column_view const& col,
    int32_t begin,
    int32_t end,
    Scalar const& value,
    std::size_t stream) {
  return COL(cudf::fill(col, begin, end, value.inner(), S(stream)));
}

std::unique_ptr<Table> repeat_table(
    Table const& tbl,
    int32_t count,
    std::size_t stream) {
  return TBL(cudf::repeat(tbl.cached_view(), count, S(stream)));
}

std::unique_ptr<Column> sequence_column(
    int32_t count,
    Scalar const& init,
    Scalar const& step,
    std::size_t stream) {
  return COL(cudf::sequence(count, init.inner(), step.inner(), S(stream)));
}

std::unique_ptr<Table> repeat_table_column(Table const& tbl, cudf::column_view const& counts, std::size_t stream) {
  return TBL(cudf::repeat(tbl.cached_view(), counts, S(stream)));
}

std::unique_ptr<Column> calendrical_month_sequence(int32_t count, Scalar const& init, int32_t months, std::size_t stream) {
  return COL(cudf::calendrical_month_sequence(count, init.inner(), months, S(stream)));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/filling.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Column = crate::ffi::Column;
        type Table = crate::ffi::Table;
        type Scalar = crate::ffi::Scalar;
        #[namespace = "cudf"]
        type column_view = crate::ffi::column_view;

        // -- Fill operations --

        /// Fills a range [begin, end) in a column with a scalar value.
        fn fill_column(
            col: &column_view,
            begin: i32,
            end: i32,
            value: &Scalar,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Repeats each row of a table N times.
        fn repeat_table(tbl: &Table, count: i32, stream: usize) -> Result<UniquePtr<Table>>;

        /// Generates an arithmetic sequence column.
        fn sequence_column(
            count: i32,
            init: &Scalar,
            step: &Scalar,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        // -- Fill (new) --

        /// Repeat table rows by per-row counts in a column.
        fn repeat_table_column(
            tbl: &Table,
            counts: &column_view,
            stream: usize,
        ) -> Result<UniquePtr<Table>>;
        /// Generate calendrical month sequence.
        fn calendrical_month_sequence(
            count: i32,
            init: &Scalar,
            months: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;
    }
}
